### PR TITLE
CXL like or dislike

### PR DESCRIPTION
**Pull request by dominik-stypula-polcode**

### CODE REVIEW

File: packages/cxl-ui/src/components/cxl-like-or-dislike.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+   */
+  @property({ type: String })
+  apiUrl = 'https://jsonplaceholder.typicode.com/posts';
+
+  static get styles() {
+    return [cxlLikeOrDislikeStyles];
+  }
+
+  async _upVote(event) {
+    this.value = this.value === 1 ? 0 : 1;
```
Comment (anoblet) [inline packages/cxl-ui/src/components/cxl-like-or-dislike.js:47]: This seems to only toggle between 0 and 1.

File: packages/cxl-ui/src/components/cxl-like-or-dislike.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    await this._vote(event.currentTarget);
+  }
+
+  async _downVote(event) {
+    this.value = this.value === -1 ? 0 : -1;
```
Comment (anoblet) [inline packages/cxl-ui/src/components/cxl-like-or-dislike.js:52]: Same thing, this seems to only toggle between 0 and -1.

**Comment by anoblet** [comment]
Is the strategy here to keep a running total? Or do we want to track upvotes and downvotes separately?

**Comment by dominik-stypula-polcode** [comment]
> Is the strategy here to keep a running total? Or do we want to track upvotes and downvotes separately?

we count only upvotes

File: packages/cxl-ui/src/components/cxl-like-or-dislike.js (hunk 1 above)
Comment (dominik-stypula-polcode) [reply packages/cxl-ui/src/components/cxl-like-or-dislike.js:47]: what is your suggestion here?

Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-like-or-dislike.js:47]: Sorry, I may have jumped the gun. Let me spend some more time to understand everything :)

**Comment by anoblet** [comment]
I see what's happening now, the 1/-1 is for tracking state for that user only.

File: packages/cxl-ui/src/components/cxl-like-or-dislike.js (hunk 1 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-like-or-dislike.js:47]: I see what's happening now. It's tracking state for the unique user only.

File: packages/cxl-ui/src/components/cxl-like-or-dislike.js (hunk 2 above)
Comment (anoblet) [reply packages/cxl-ui/src/components/cxl-like-or-dislike.js:52]: Disregard

**Comment by anoblet** [comment]
One thing we might consider doing is saving the user's upvote/downvote on the server-side rather than in local storage.

The reason for this being that if the user decides to clear their cookies/cache for whatever reason, there won't be a inconsistent state.

**Comment by pawelkmpt** [comment]
@dominik-stypula-polcode I think this component does not display values correctly (or I don't know how to use it).


`$upvotes` carries ~sum~ difference of all votes (`upvotes - downvotes`).

```html
<cxl-like-or-dislike
    apiUrl="$url"
    postType="$post->post_type"
    postId="$post->ID"
    userId="$user_id"
    upVotes="$upvotes"
></cxl-like-or-dislike>
```

I have a case when playbook has:
* upvotes: 0
* downvotes: 1

what gives: `-1` but component shows `-2`. I display page as user who voted.
For a different user (who didn't vote) it displays: `-1`.

Approach 2: I pass number of upvotes (not sum like previously).
* user who voted sees: `-1`
* other user sees: `0`

**Comment by pawelkmpt** [comment]
I'm also wondering if we should have test endpoint hardcoded?

https://github.com/dominik-stypula-polcode/aybolit/blob/ec1e40ec094cb9a69278d676322fd1a612858bb4/packages/cxl-ui/src/components/cxl-like-or-dislike.js#L34-L40

**Comment by dominik-stypula-polcode** [comment]
> 
> I have a case when playbook has:
> 
> * upvotes: 0
> * downvotes: 1
> 
> what gives: `-1` but component shows `-2`. I display page as user who voted.
> For a different user (who didn't vote) it displays: `-1`.

@pawelkmpt you're right, the sum of votes + user vote was in the wrong place in the code.
I fixed that now. 
Please check and confirm if it's ok now.

Bit of explanation: when user votes, the component makes an API call which should register the vote in the database, besides that component registers the vote (-1 or +1) in the browser's `localStorage` (see Console->Application->Local Storage) to check the correct thumb-up/thumb-down icon after the page is refreshed so user knows he already voted for the playbook without a need to ask database for that (yet we could add an additional attribute like `userVote` which could fetch the specific user's vote, but we tried to make it simple).
User can also uncheck his vote so in this case we make an API call with "0" value to confirm that user has resigned of his vote (no icon should be checked)

**Comment by dominik-stypula-polcode** [comment]
> I'm also wondering if we should have test endpoint hardcoded?

Not necessarily, but if it's empty you'll get an error.

**Comment by pawelkmpt** [comment]
New PR and merge here: https://github.com/conversionxl/aybolit/pull/166

**Comment by pawelkmpt** [comment]
@dominik-stypula-polcode I think you can close this PR. I don't have such button available

**Pull request merged**